### PR TITLE
Split "/company/register_productive_consumption"

### PR DESCRIPTION
**Pull request by sloschert**
This commit splits the route "/company/register_productive_consumption" [GET, POST], into two seperate routes, "select_productive_consumption" (GET) and "register_productive_consumption" (POST).

This is done in preparation of the next steps laid down in "RFC: Change productive consumption user flow" (#1053).

Plan: 371825c0-3112-47cd-9cb0-fea799fad3ee (5x)

### CODE REVIEW

**Comment by Amittai-Aviram** [comment]
> I think that there is a problem with the redirection logic in this PR. 

But is that not the desired behavior?  Would you expect a consumption value to be negative?

**Comment by sloschert** [comment]
> I would expect to be presented with a form that allows me to correct the incorrect amount field.

I'm on it.

**Review by Amittai-Aviram — COMMENTED** [review]
The redirect logic needs to be corrected, as we agreed.  Otherwise, this looks good!

**Review by Amittai-Aviram — COMMENTED** [review]
The redirect logic needs to be corrected, as we have agreed.  Otherwise, this looks good!

**Comment by 4lm** [comment]
Thank you for creating this PR. I think, I now understand more clearly, what you aim to achieve with this PR. However, as I said in #1053, I don’t think it’s necessary to split the route. Also, as it turns out (something I didn’t realize earlier), we already have a plan selection route at /user/query_plans (labeled “All plans”).

When a user clicks on a plan there ("All plans"), they are taken to the corresponding detail page with a parameterized link to the registration of consumption page. The "All plans" route also allows users to search by plan name and plan ID, making it effective for plan selection. Therefore, there is no need for an additional selection route.

By using HTTP request methods (GET or POST) on the `/register_productive_consumption` route for control flow, we can avoid the issues that arose in this PR. Using HTTP verbs like GET and POST in a more semantic way, helps with control flow clarity, and aligns with best practices in web development. HTTP verbs are fundamental to REST (Representational State Transfer) architecture, which is a widely adopted pattern for designing networked applications, including web apps serving HTML.

The following pseudo-code demonstrates how all cases can be handled solely with the `/register_productive_consumption route`:

```
IF request method is POST:
    retrieve form data for 'plan', 'amount', 'consumption_type'
    validate form data independently
    IF any validation fails:
        flash error messages
        render form with pre-filled valid values
    ELSE:
        try:
            perform transaction
        except Exception as e:
            flash error message
            render form with pre-filled valid values
        ELSE:
            flash success message
            redirect (GET, 302) to the same route to clear the form

ELSE (request method is GET):
    retrieve query parameters for 'plan', 'amount', 'consumption_type'
    IF any query parameters are present:
        render form with provided parameters
    ELSE:
        render an empty form
```

Further, renaming the “All plans” route to something more descriptive and giving it a more prominent placement might be beneficial. Additionally, adding a parameterized link next to each plan in the list of this route would help users navigate directly to the productive consumption registration page.

**Comment by 4lm** [comment]
To be fair. Even though I think one route would be more clean, to either have one or two routes is not really the important part. The important part is to use HTTP methods (verbs) in a semantic way. Therefore, specific to this PR, you also could use two routes, with one verb each. Which verb? The one that reflects the semantic meaning of the specific route. More clearly:

- only allow GET on `/select_productive_consumption`
- only allow POST on `/register_productive_consumption`

To fix the above (by seppljordan) raised issue is quite easy then. You only would have to change the behavior of the two routes in this PR only so slightly to something as following.

GET route `/select_productive_consumption`:

Please note: The only changes this route needs is (i) a restriction to using GET, the form already points correctly to `/register_productive_consumption` via a `<form>`-POST method and  (ii) to recognize the `amount` and `type_of_consumption` params in an URL query path, to pre-fill the form, if present.

```
IF request method is GET:
    retrieve query parameters for 'plan', 'amount', 'consumption_type'
    IF any query parameters are present:
        render form with provided parameters
    ELSE:
        render an empty form
```

POST route `/register_productive_consumption`:

```
IF request method is POST:
    retrieve form data for 'plan', 'amount', 'consumption_type'
    validate form data independently
    IF any validation fails:
        flash error messages
        redirect (GET, 302) to `/select_productive_consumption...` with params in URL path (if applicable), pre-fills form
    ELSE:
        try:
            perform transaction
        except Exception as e:
            flash error message
            redirect (GET, 302) to `/select_productive_consumption...` with params in URL path (if applicable), pre-fills form
        ELSE:
            flash success message
            redirect (GET, 302) to `/select_productive_consumption` with no params (results is empty form)
```

File: arbeitszeit_flask/company/routes.py [hunk 1]
```diff
@@ -224,7 +227,12 @@ def hide_plan(plan_id: UUID, hide_plan: HidePlan, presenter: HidePlanPresenter):
 class register_hours_worked(RegisterHoursWorkedView): ...
 
 
-@CompanyRoute("/register_productive_consumption", methods=["GET", "POST"])
+@CompanyRoute("/select_productive_consumption")
```
[the remark this answers is not in the document]
Comment (sloschert) [reply arbeitszeit_flask/company/routes.py:230]: yes, it defaults to GET, that's flask behaviour. I'm not sure if it would be cleaner to specify it.

**Comment by sloschert** [comment]
> IF any validation fails:
>         flash error messages

This part is actually a bit complicated because the `RegisterProductiveConsumptionController` currently attaches errors to form fields. But it's not possible to send a form back to the "consumption selection" GET route, so this error attachment system has to be replaced with flash messaging. On top of that, we seem to have some problems with the flash messaging system - messages do not show up after a redirect, I'll have to investigate where this comes from. 

So this change will probably take a bit longer...

**Comment by sloschert** [comment]
Closing this PR. A split of this route is probably not necessary. I'll start over.

**Pull request closed without merging**